src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -89,18 +90,38 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function compile()
     {
+        if ($this->isCompiled()) {
+            return;
+        }
+
         $this->parameterBag->resolve();
 
-        $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        if (!$this->parameterBag instanceof FrozenParameterBag) {
+            $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        }
+
+        $this->compiled = true;
     }
 
     /**
-     * Returns true if the container parameter bag are frozen.
+     * Returns true if the container is compiled.
      *
-     * @return bool true if the container parameter bag are frozen, false otherwise
+     * @return bool
+     */
+    final public function isCompiled()
```

Comment:
This is now the ought to use method. If you truely want to check for a frozen parameter bag use `getParameterBag() instanceof FrozenParameterBag` instead (hence the deprecation).

Besides that this method guarantees state of `compile()` as we dont depend on a frozen parameterbag for that anymore (which we can pass to the constructor ;-)).